ra/ra.go

```diff
@@ -712,13 +712,23 @@ func (ra *RegistrationAuthorityImpl) checkOrderAuthorizations(
 	if err != nil {
 		return nil, err
 	}
+
 	// Ensure the names from the CSR are free of duplicates & lowercased.
 	names = core.UniqueLowerNames(names)
+
 	// Check the authorizations to ensure validity for the names required.
 	if err = ra.checkAuthorizationsCAA(ctx, names, authzs, ra.clk.Now()); err != nil {
 		return nil, err
 	}
 
+	// Check the challenges themselves too.
+	for _, authz := range authzs {
+		chall, err := authz.SolvedBy()
+		if err != nil || chall == "" || (authz.Wildcard && chall != core.ChallengeTypeDNS01) {
+			return nil, errors.New("authorization fulfilled by invalid challenge")
+		}
+	}
+
```

Comment:
Good points! I've split this out into separate checks so that we can return different error messages for each.

I've also taken it a step further. Your point about using the PA is a good one, although we can't use checkOrderNames and WillingToIssueWildcards for this -- that only takes into account whether the requested identifier is a wildcard name, not which challenge we ended up using for that name. But I'm able to use the PA's `ChallengesFor` function to get the list of challenges we'd consider acceptable for the identifier, and then make sure that the actual challenge used is contained in that list.

I don't love this solution, as ChallengesFor actually _constructs_ new challenges, not just lists their types. I'd love more feedback on this approach.

Reply:
>  although we can't use checkOrderNames and WillingToIssueWildcards for this -- that only takes into account whether the requested identifier is a wildcard name, not which challenge we ended up using for that name.

Sure, I get that. I think I was attempting to use the fact that the RA makes a call to the PA a little while later as an illustrative purpose.

> I don't love this solution, as ChallengesFor actually constructs new challenges, not just lists their types. 
That happens because ChallengesFor runs `	token := core.NewToken()`? Yeah. This is wasteful because if we're just doing a check, we shouldn't need to generate this random token. It seems to me that ChallengesFor is fulfilling two separate roles
1. It's determining what challenges are acceptable for the identifier (copypasted from the comment about the function)
2. It creates challenges for each type of enabled challenge. I think this role might be able to be factored out into a separate function, something like

```
// ChallengesFor makes a decision of what challenges are acceptable for
// the given identifier.
func (pa *AuthorityImpl) CheckChallengesFor(identifier identifier.ACMEIdentifier) ([]challenges, error) {
	challenges := []string

	// If the identifier is for a DNS wildcard name we only
	// provide a DNS-01 challenge as a matter of CA policy.
	if strings.HasPrefix(identifier.Value, "*.") {
		// We must have the DNS-01 challenge type enabled to create challenges for
		// a wildcard identifier per LE policy.
		if !pa.ChallengeTypeEnabled(core.ChallengeTypeDNS01) {
			return nil, fmt.Errorf(
				"Challenges requested for wildcard identifier but DNS-01 " +
					"challenge type is not enabled")
		}
		// Only provide a DNS-01-Wildcard challenge
		challenges = append(core.ChallengeTypeDNS01)
	} else {
		// Otherwise we collect up challenges based on what is enabled.
		if pa.ChallengeTypeEnabled(core.ChallengeTypeHTTP01) {
			challenges = append(core.ChallengeTypeHTTP01)
		}

		if pa.ChallengeTypeEnabled(core.ChallengeTypeTLSALPN01) {
			challenges = append(core.ChallengeTypeTLSALPN01)
		}

		if pa.ChallengeTypeEnabled(core.ChallengeTypeDNS01) {
			challenges = append(core.ChallengeTypeDNS01)
		}
	}

	return challenges, nil
}

// CreateChallenges generates challenges for acceptable identifiers
func (pa *AuthorityImpl) CreateChallenges(identifier identifier.ACMEIdentifier, acceptableChallenges []challenges) ([]core.Challenge, error) {
	challenges := []core.Challenge{}

	token := core.NewToken()
    for chall := range acceptableChallenges {
    	challenges = append(challenges, core.<__ChallengeTypePassedIntoThisFunction__>(token))
	}

	// We shuffle the challenges to prevent ACME clients from relying on the
	// specific order that boulder returns them in.
	shuffled := make([]core.Challenge, len(challenges))

	pa.rngMu.Lock()
	defer pa.rngMu.Unlock()
	for i, challIdx := range pa.pseudoRNG.Perm(len(challenges)) {
		shuffled[i] = challenges[challIdx]
	}

	return shuffled, nil
}
```

Reply:
That's quite a bit bigger change though.

Reply:
Yep, I had the same thought, about breaking that function into a "just give me the challenge types" function and a "actually create the challenges" function which relies on the first one to do its work. I was calling them `ChallengeTypesFor` and `ChallengesFor` in my own mind, but the idea is the same.

The difficulty is this line in your proposed change:
```go
    	challenges = append(challenges, core.<__ChallengeTypePassedIntoThisFunction__>(token))
```

The `CheckChallengesFor` / `ChallengeTypesFor` function would be returning a slice of `core.AcmeChallenge`. The `CreateChallengesFor` / `ChallengesFor` function would be returning a slice of `core.Challenge`. But `core.AcmeChallenge` isn't a constructor -- it's just a type-safe string with the contents "DNS-01" or "HTTP-01" or whatever. We need a way to map a concrete value of `core.AcmeChallenge` (e.g. `core.ChallengeTypeHTTP01`) to a concrete one of the Challenge constructors (e.g. `core.HTTPChallenge01()`).

Of course, this is a solvable problem, but it does expand the scope of the change beyond the PA and into the core library, so at that point we have to ask ourselves "is this worth it?" To which the answer very well may be yes! But I'd like more input.

Reply:
The condition we are trying to ensure is that authorizations for wildcard identifiers (`*.example.com`) can only be validated by DNS-01 challenges.

We ensure this condition at authz creation time, but it's reasonable to want to re-check this condition each time we rehydrate an authz from the database, or (in this case) right before issuance.

Given that we want to ensure the condition separately at creation time and at reading time, it's natural we'll need two different methods of ensuring it. We already have the creation-time method (in PolicyAuthority.ChallengesFor) and we are adding a read-time method here, in a somewhat ad-hoc manner.

To make the read-time method less ad-hoc, it makes sense to put it in the `policy` package alongside `ChallengesFor`. I think it should take the form of `AuthzValid(*core.Authorization) error`, which checks this specific constraint (if authz.Wildcard && solved by non-DNS-01, error).

Also FYI @pgporada, per RFC 8555 the identifier field never contains a `*`. That's why we have the `wildcard` boolean field in authorization objects. So an order for `*.example.com` would result in an authz for `example.com` with `"wildcard": true` in the JSON.

Reply:
I have a demo of what this would look like here: https://github.com/letsencrypt/boulder/compare/finalize-refactor...finalize-refactor-challs-for?expand=1

Reply:
Okay, I've updated this PR as per both Jacob and Phil's feedback. PTAL.